nutils/evaluable.py

```diff
@@ -3900,6 +3900,175 @@ def _intbounds_impl(self):
+        The source coordinate system.
+    index : scalar, integer :class:`Array`
+        The index part of the source coordinates.
+    '''
+
+    def __init__(self, target, source, index: Array):
+        if index.dtype != int or index.ndim != 0:
+            raise ValueError('argument `index` must be a scalar, integer `nutils.evaluable.Array`')
+        self._target = target
+        self._source = source
+        self._index = index
+        super().__init__(args=[index], shape=(), dtype=int)
+
+    def evalf(self, index):
+        if self._target is not None:
+            index, _ = self._target.index_with_tail(self._source[index.__index__()])
+        else:
+            index = 0
+        return numpy.array(index)
+
+    def _intbounds_impl(self):
+        len_target = 1 if self._target is None else len(self._target)
+        return 0, len_target - 1
+
+    def _simplified(self):
+        if self._target is None:
+            return ones((1,), dtype=int)
+        elif self._target == self._source:
+            return self._index
+
+
+class TransformLinear(Array):
+    '''Linear part of a coordinate transformation
+
+    Args
+    ----
+    target : :class:`nutils.transformseq.Transforms`, optional
+        The target coordinate system. If `None` the target is 'the' reference
+        coordinate system.
+    source : :class:`nutils.transformseq.Transforms`
+        The source coordinate system.
+    index : scalar, integer :class:`Array`
+        The index part of the source coordinates.
+    '''
+
+    def __init__(self, target, source, index: Array):
+        if index.dtype != int or index.ndim != 0:
+            raise ValueError('argument `index` must be a scalar, integer `nutils.evaluable.Array`')
+        self._target = target
+        self._source = source
+        target_dim = source.todims if target is None else target.fromdims
+        super().__init__(args=[index], shape=(target_dim, source.fromdims), dtype=float)
+
+    def evalf(self, index):
+        chain = self._source[index.__index__()]
+        if self._target is not None:
+            _, chain = self._target.index_with_tail(chain)
+        if chain:
+            return functools.reduce(lambda r, i: i @ r, (item.linear for item in reversed(chain)))
+        else:
+            return numpy.eye(self._source.fromdims)
+
+    def _simplified(self):
+        if self._target == self._source:
+            return diagonalize(ones((self._source.fromdims,), dtype=float))
+
+
+class TransformBasis(Array):
+    '''Extended basis of a source coordinate system in reference coordinates
+
+    The basis (the columns of this matrix) spans the space of reference
+    coordinates (spatial part). If the source coordinate system has a lower
+    dimension, say `n`, than the reference coordinate system, the first `n`
+    basis vectors span the space of source coordinates mapped to the reference
+    for the given source index. The remainder is *not* a span of the complement
+    space.
```

Comment:
I think I would understand this better if it was introduced along the lines of "given a transformation chain into n-dimensional space, the TransformBasis represents a vector basis of dimension n with the special property that columns 1 .. m < n form a basis for any m-dimensional subspace traversed by the transformation chain". I'm also not sure about calling it an "Extended basis" which is not very informative. Perhaps "cascaded vector basis for a transformation chain" connects better with the suggested paragraph.

Reply:
Let's drop extended indeed. I'm not sure if we should mention transformation chains. That's an implementation detail. I'd prefer to talk about coordinate systems and mappings from one to another. Let's discuss this next week.